nvflare/app_common/resource_managers/list_resource_manager.py

```diff
@@ -12,66 +12,138 @@
         Args:
             resources (dict): Specify the list of resources unit
+            expiration_period (int): Number of seconds to hold the resources reserved.
+                If check_resources is called but after "expiration_period" no allocate resource is called,
+                then the reserved resources will be released.
         """
         super().__init__()
         if not isinstance(resources, dict):
             raise TypeError(f"resources should be of type dict, but got {type(resources)}.")
+        if not isinstance(expiration_period, int):
+            raise TypeError(f"expiration_period should be of type int, but got {type(expiration_period)}.")
+        if expiration_period < 0:
+            raise ValueError("expiration_period should be greater than 0.")
+
         self.resources = {}
         for k in resources:
             if not isinstance(resources[k], list):
                 raise TypeError(f"item in resources should be of type list, but got {type(resources[k])}.")
-            self.resources[k] = set(resources[k])
+            self.resources[k] = deque(list(dict.fromkeys(resources[k])))
+
+        self.expiration_period = expiration_period
+        self.reserved_resources = {}
+        self.lock = Lock()
+        self.stop_event = Event()
+        self.cleanup_thread = Thread(target=self._check_expired)
+
+    def _check_expired(self):
+        while not self.stop_event.is_set():
+            time.sleep(1)
+            with self.lock:
+                tokens_to_remove = []
+                for k in self.reserved_resources:
+                    r, t = self.reserved_resources[k]
+                    t -= 1
+                    if t == 0:
+                        tokens_to_remove.append(k)
+                    else:
+                        self.reserved_resources[k] = r, t
+                for token in tokens_to_remove:
+                    reserved_resources, _ = self.reserved_resources.pop(token)
+                    for k in reserved_resources:
+                        for i in reserved_resources[k]:
+                            self.resources[k].append(i)
+
+    def handle_event(self, event_type: str, fl_ctx: FLContext):
+        if event_type == EventType.SYSTEM_START:
+            self.cleanup_thread.start()
+        elif event_type == EventType.SYSTEM_END:
+            self.stop_event.set()
+            if self.cleanup_thread:
+                self.cleanup_thread.join()
+                self.cleanup_thread = None
 
     def check_resources(self, resource_requirement: dict, fl_ctx: FLContext) -> (bool, Optional[str]):
         if not isinstance(resource_requirement, dict):
             raise TypeError(f"resource_requirement should be of type dict, but got {type(resource_requirement)}.")
+
         check_result = True
-        for k in resource_requirement:
-            if k in self.resources:
-                if len(self.resources[k]) < resource_requirement[k]:
+        token = None
+        with self.lock:
```

Comment:
L100 - L110, 
1)   should we refactor this code into a function that can easily add unit test to test logic
for example: 
def   check_available_resources (resource_requirement : dict)  -> Boolean : 
        ...
        return check_result
2) check_result seems to be a "global" flag that covers all type of resource requirements
      suppose we have two types of resources : gpu, cpu
      if the gpu is not available check_results = False, then cpu will get the same Check_result = False. 
      Is this intended ?  should the check_result be a dictionary, that express something like
           check_result['gpu'] = False
           check_result['cpu'] = True 
? 

if this make sense, the above function signature would be 
def   check_available_resources (resource_requirement : dict)  -> dict : 

and the overall function signature : probably need to change to 
def check_resources(self, resource_requirement: dict, fl_ctx: FLContext) -> (dict, Optional[str]):

Reply:
1. => sure I can do that.

2. => Yes, it is intended, you need to have ALL the resource available for the check to pass.
=> it depends on what we want to do with this ListResourceManager, since the user can write their own ResourceManager, I prefer to keep this class perform this way for simplicity

Reply:
Ok, make sense